crates/rpc/rpc/src/eth/gas_oracle.rs

```diff
@@ -0,0 +1,264 @@
+        }
+
+        // if all responses are empty, then we can return a maximum of 2*check_block blocks' worth
+        // of prices
+        //
+        // we only return more than check_block blocks' worth of prices if one or more return empty
+        // transactions
+        let mut current_block = header.number;
+        let mut results = Vec::new();
+        let mut populated_blocks = 0;
+
+        // we only check a maximum of 2 * max_block_history
+        for _ in 0..self.oracle_config.max_block_history * 2 {
+            let block_values = self
+                .get_block_values(current_block, SAMPLE_NUMBER as usize)
+                .await?
+                .ok_or(GeneralError::Provider(ProviderError::BlockBodyIndices {
+                    number: current_block,
+                }))?;
+
+            if block_values.is_empty() {
+                results.push(U256::from(last_price.price));
+            } else {
+                results.extend(block_values);
+                populated_blocks += 1;
+            }
+
+            // break when we have enough populated blocks
+            if populated_blocks >= self.oracle_config.blocks {
+                break
+            }
+
+            current_block -= 1;
+        }
+
+        // sort results then take the configured percentile result
+        let mut price = last_price.price;
+        if !results.is_empty() {
+            results.sort_unstable();
+            price = *results
+                .get((results.len() - 1) * self.oracle_config.percentile as usize / 100)
+                .expect("gas price index is a percent of nonzero array length, so a value always exists; qed");
+        }
+
+        // constrain to the max price
+        if let Some(max_price) = self.oracle_config.max_price {
+            if price > max_price {
+                price = max_price;
+            }
+        }
+
+        *last_price = GasPriceOracleResult { block_hash: header_hash, price };
+
+        Ok(price)
+    }
+
+    /// Get the `limit` lowest effective tip values for the block at number `block_num`. If the
+    /// oracle has a configured `ignore_price` threshold, then tip values under that threshold will
+    /// be ignored before returning a result.
+    ///
+    /// If the block cannot be found, then this will return `None`.
+    async fn get_block_values(&self, block_num: u64, limit: usize) -> Result<Option<Vec<U256>>> {
+        // TODO: we could cache num -> hash as well as long as we invalidate the cache between
+        // forkchoice updates
+        let block_hash = match self.client.block_hash(block_num)? {
+            Some(num) => num,
+            None => return Ok(None),
+        };
+
+        // check the cache
+        let block = match self.cache.get_block(block_hash).await? {
+            Some(block) => block,
+            None => match self.client.block_by_number(block_num)? {
+                Some(block) => block,
+                None => return Ok(None),
+            },
```

Comment:
this is redundant, because the cache service would have found the block if it exists.

Reply:
no longer necessary since the move to traversing by parent hash